### PR TITLE
Constructors

### LINKED ISSUE #16: implement multiple constructors, so more than just vector may be passed into Collection

### PR DESCRIPTION
**Pull request by ianzapolsky**
- [x] std::list constructor
- [x] std::array constructor
- [x] c-style array constructor

closes #16

### CODE REVIEW

File: src/collections.h [hunk 1]
```diff
@@ -12,10 +15,29 @@ class Collection {
 private:
     std::vector<T> Data;
 public:
+    // std::vector constructor
     Collection<T>(std::vector<T> d) {
         Data = d;
     }
 
+    // std::list constructor
+    Collection<T>(std::list<T> d) {
```
Comment (natebrennand) [inline src/collections.h]: `size()` is a constant time operation so we should instantiate `Data` to the correct size.

File: src/collections.h [hunk 2]
```diff
@@ -12,10 +15,29 @@ class Collection {
...
+        for (auto i : d)
+            Data.push_back(i);       
+    }
+
+    // std::array constructor
+    template<size_t N>
+    Collection<T>(std::array<T, N> d) {
+        for (auto i : d)
```
Comment (natebrennand) [inline src/collections.h]: same here

**Comment by natebrennand** [comment]
Should we include transformations back to vector/list/array in this PR?

**Comment by ianzapolsky** [comment]
makes sense, though i'm unclear how the syntax of those transformations should look.

**Comment by ianzapolsky** [comment]
`toVector`, `toList`, `toArray`?

**Comment by natebrennand** [comment]
Ya, sgtm. Should we even include "to" in the name?

Or is it better for us to expose all the iteration methods that would allow using `Collection` as a constructor argument? I think it's called the "allocator trait" or something like that?

**Comment by ianzapolsky** [comment]
so we're talking about this type of difference:

```
std::list<T> mylist = Collection<T>::range(1,11).list();
```
vs.
```
std::list<T> mylist = std::list<T>(Collection<T>::range(1,11));
```

**Comment by natebrennand** [comment]
Exactly, I'm not sure which is better to have. I think we should have the latter no matter what.
Then we just need to decide if we want to include the former as wel.

**Comment by ianzapolsky** [comment]
I'm trying to do some research on allocator_trait and not finding much. It does not look like you can construct lists from vectors and vice versa, you can only do so by defining a range over the collection...

**Comment by ianzapolsky** [comment]
e.g.
```
std::list<int> l;
std::vector<T> v{ std::begin(l), std::end(l) };
```

**Comment by natebrennand** [comment]
So it looks like you kinda can: http://stackoverflow.com/a/5218738/1763664
but that it's not made very easy. I think we should probably just stick with the `toX` methods in that case.

**Comment by ianzapolsky** [comment]
so i'm running into a problem with implementing `toX` method for `std::array`. Basically, I think it's impossible to have a function that returns a std::array whose length is unknown at compile time. I cannot figure out how to get the templates to work around this big problem. see the latest definition of the array() function on this branch:

**Comment by natebrennand** [comment]
I don't think it's a huge problem if we can't get `toArray`. There aren't a ton of use cases for arrays as opposed to vectors anyways.

**Comment by ianzapolsky** [comment]
that's exactly my thinking. the fixed length of arrays doesn't really gel with the point of the lib. anyway

**Comment by natebrennand** [comment]
let's just scrap the `toArray` transformer and then we're good to go

**Comment by ianzapolsky** [comment]
btw, I opted to name them simply vector and list. open to calling them `toVector` and `toList` though..

**Comment by natebrennand** [comment]
I'm indifferent on the names. Both seem plenty descriptive

**Comment by natebrennand** [comment]
:+1:  to merger

**Pull request merged**